pkg/storage/bloom/v1/ast_extractor.go

```diff
@@ -27,37 +27,67 @@ type OrLabelMatcher struct{ Left, Right LabelMatcher }
 // if both of the Left and Right label matcher bloom tests pass.
 type AndLabelMatcher struct{ Left, Right LabelMatcher }
 
+type seriesLabelNames map[string]struct{}
+
+func newSeriesLabelNames(series []labels.Labels) seriesLabelNames {
+	lblNames := make(seriesLabelNames, len(series))
+	for _, lbls := range series {
+		for _, lbl := range lbls {
+			lblNames[lbl.Name] = struct{}{}
+		}
+	}
+	return lblNames
+}
+
+func (s seriesLabelNames) contains(name string) bool {
+	_, ok := s[name]
+	return ok
+}
+
 // ExtractTestableLabelMatchers extracts label matchers from the label filters
 // in an expression. The resulting label matchers can then be used for testing
 // against bloom filters. Only label matchers before the first parse stage are
 // included.
+// If series are provided, labels matchers that are part of the series are not included
 //
 // Unsupported LabelFilterExprs map to an UnsupportedLabelMatcher, for which
 // bloom tests should always pass.
-func ExtractTestableLabelMatchers(expr syntax.Expr) []LabelMatcher {
+func ExtractTestableLabelMatchers(expr syntax.Expr, series ...labels.Labels) []LabelMatcher {
```

Comment:
It's _probably_ ok to have this logic here, but I've never been happy with the [cognitive overhead](https://github.com/grafana/loki/blob/4687f377db0a7ae07ffdea354582c882c10b72c4/pkg/storage/bloom/v1/bloom_tester.go#L164-L172) in `bloom_tester.go` and I suspect it might be easier to follow if we moved the series knowledge checks down to the bloom test itself. 

WDYT? 

If we do move the logic that far down, we'll need to update other functions to make sure the labels get propagated through properly, like inside `partitionRequest`.

Reply:
Done. This way we can also filter out
```
{app="fake"} | app="noexist"
{app="fake"} | app=""
```